### DISCUSSION TITLE
What am I doing wrong? Asyncio

### DISCUSSION THREAD

**Discussion by limzyishere** [python-telegram-bot/python-telegram-bot#4108] [Q&A]
Hello! I tried to make it as easy to digest as possible. I am a beginner in both PTB and Asyncio so any help would be much appreciated.

File A
```python
scheduler = BackgroundScheduler()

 #Code abbreviated for simplicity.
async def daily_sale_summary():
           for x in y:
            task = tele_daily_summary_msg( x)
            tasks.append(task)
            
await asyncio.gather(*tasks)


def run():
    asyncio.run(daily_sale_summary())


scheduler.add_job(run, "interval", seconds=5)
scheduler.start()
```

File B
```python
bot = Bot(BOT_TOKEN)

async def tele_daily_summary_msg(x):

    await bot.send_message(
        chat_id=telegram_chat_id,
        text=f"{x}"
        message_thread_id=55,
        parse_mode=ParseMode.HTML,
    )
```

Running File A will result in 

Issue A
> telegram.error.TimedOut: Pool timeout: All connections in the connection pool are occupied. Request was *not* sent to Telegram. Consider adjusting the connection pool size or the pool timeout.

As I am using Bot, there is no option to adjust the connection pool size, ok so I have extracted from File C my Application object and added connection pool and pool timeout. 

File C
```python

def create_telegram_bot():
    application = (
        Application.builder()
        .connection_pool_size(5000)
        .pool_timeout(5000)
        .token(BOT_TOKEN)
        .build()
    )
    application.add_handler(conv_handler)
    application.add_handler(get_topic_handler)

    return application
```

This eliminates the above error. However I get a new error -

Issue B
> telegram.error.NetworkError: Unknown error in HTTP implementation: RuntimeError('Event loop is closed')

I understand this has something to do with using threading that causes a new event loop to spawn, but where am I calling a new thread?

Also, how do I determine what is a good connection_pool_size and pool_timeout? I don't understand why Issue A happened when my asyncio.gather only contained 2 tasks.

**Comment by Bibo-Joshi** [comment]
> As I am using Bot, there is no option to adjust the connection pool size

You can very well adjust that. the `Bot` class has a parameter [`request`](https://docs.python-telegram-bot.org/en/stable/telegram.bot.html#telegram.Bot.params.request) to which you can pass an instance of [`HTTPXRequest`](https://docs.python-telegram-bot.org/en/stable/telegram.request.httpxrequest.html) with the desired configuration.

side notes:

* apscheduler has an asyncio-friendly scheduler built-in
* PTB has the built-in [`JobQueue`](https://github.com/python-telegram-bot/python-telegram-bot/wiki/Extensions---JobQueue) which might be useful for you as well (and is in fact built with APScheduler)
